Fix a bug #3994

DjangoPeng opened the pull request:
I have fixed the bug[#3994](https://github.com/tensorflow/tensorflow/issues/3994). So, now, we can use `Variables` as gradients with the update.
Check if gradient is `tf.Variable`:
- Yes, convert it to tensor.
- No, check its type as before.

Then, remake the `grads_and_vars` after checking and converting.

DjangoPeng committed:
Fix a bug #3994

tensorflow-jenkins commented:
Can one of the admins verify this patch?

danmane commented:
@mrry Can you review this please?

mrry commented on tensorflow/python/training/optimizer.py:
```diff
@@ -27,7 +27,7 @@
 from tensorflow.python.ops import variables
 from tensorflow.python.training import slot_creator
-
+import tensorflow as tf
```
Remove this import (as it creates a circular dependency, and the relevant method is available from `ops`).

mrry commented on tensorflow/python/training/optimizer.py:
```diff
@@ -282,8 +282,14 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
+
+    grads_to_tensors_list = []
+    vars_list = []
     grads_and_vars = tuple(grads_and_vars)  # Make sure repeat iteration works
```
I think this line is redundant now. (It was previously necessary in order to be able to iterate over the argument twice, when the argument was a generator expression. However, you now only iterate over it once.)

mrry commented on tensorflow/python/training/optimizer.py:
```diff
@@ -282,8 +282,14 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
     for g, v in grads_and_vars:
+      # Check if gradient is Variable. If so, convert it to tensor
+      if isinstance(g, (variables.Variable)):
+        g = tf.convert_to_tensor(g)
```
It seems like it would be simpler (and cover more cases) to do:

```python
if g is not None:
  try:
    g = ops.convert_to_tensor_or_indexed_slices(g)
  except TypeError:
    raise TypeError(
        "Gradient must be convertible to a Tensor or IndexedSlices, or None: %s" %g)
```

mrry commented on tensorflow/python/training/optimizer.py:
```diff
@@ -292,6 +298,13 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
         self._assert_valid_dtypes([g, v])
+      # If all is well, append g and v to corresponding list separately 
+      grads_to_tensors_list.append(g)
+      vars_list.append(v) 
```
Instead of maintaining two lists in sync, change this to maintain a single list of tuples, called e.g. `converted_grads_and_vars` (then use `converted_grads_and_vars` in place of `grads_and_vars` in the remainder of this function).

mrry commented:
Please add a test that passes a `tf.Variable` to `Optimizer.apply_gradients()` in `optimizer_test.py`.

DjangoPeng committed:
Fix a bug #3994

DjangoPeng committed:
Add a test for passing gradients as tf.Variables

DjangoPeng commented:
@mrry  I have updated the `optimizer.py` and added the test that passes a `tf.Variable` to `Optimizer.apply_gradient()` in `optimizer_test.py`. But, I added an import of `ops`, because I have to use the `ops.control_dependencies()` to assist that convert gradient to `tf.Variable`. :(

DjangoPeng commented on tensorflow/python/training/optimizer_test.py:
```diff
@@ -113,6 +114,32 @@ def testNoGradients(self):
+        # Convert gradients to tf.Variables
+        converted_grads = [tf.Variable(tf.zeros([2], dtype)) for i in grads_and_vars]
+        convert_ops = [tf.assign(converted_grads[i], gv[0]) for i,gv in enumerate(grads_and_vars)]
+        
```
Convert the gradient(tensor) to `tf.Variable`

DjangoPeng commented on tensorflow/python/training/optimizer_test.py:
```diff
@@ -113,6 +114,32 @@ def testNoGradients(self):
+        converted_grads_and_vars = list(zip(converted_grads, [var0, var1]))
+        with ops.control_dependencies(convert_ops):
+            opt_op = sgd_op.apply_gradients(converted_grads_and_vars, global_step)
+
```
Add the `ops.control_dependencies(convert_ops):` to ensure the `converted_grads` has been initialized.

mrry commented on tensorflow/python/training/optimizer_test.py:
```diff
@@ -113,6 +114,32 @@ def testNoGradients(self):
+        convert_ops = [tf.assign(converted_grads[i], gv[0]) for i,gv in enumerate(grads_and_vars)]
+        
+        converted_grads_and_vars = list(zip(converted_grads, [var0, var1]))
+        with ops.control_dependencies(convert_ops):
```
`with tf.control_dependencies(convert_ops):` should work here.

Alternatively, you could run the `convert_ops` and `opt_op` in separate steps, which might make the test clearer.

DjangoPeng replied:
Sure thing, I agree to that the separate run is better than the `control_dependencies`. I'll update it!

mrry commented on tensorflow/python/training/optimizer_test.py:
```diff
@@ -17,6 +17,7 @@
 from __future__ import absolute_import
 from __future__ import division
 from __future__ import print_function
+from tensorflow.python.framework import ops
```
This import shouldn't be necessary.

DjangoPeng committed:
Update the test for passing gradientsAsVariables

DjangoPeng commented on tensorflow/python/training/optimizer_test.py:
```diff
@@ -128,10 +127,11 @@ def testGradientsAsVariables(self):
         converted_grads_and_vars = list(zip(converted_grads, [var0, var1]))
-        with ops.control_dependencies(convert_ops):
-            opt_op = sgd_op.apply_gradients(converted_grads_and_vars, global_step)
+        opt_op = sgd_op.apply_gradients(converted_grads_and_vars, global_step)
```
Remove the `control_dependencies`

DjangoPeng commented on tensorflow/python/training/optimizer_test.py:
```diff
@@ -128,10 +127,11 @@ def testGradientsAsVariables(self):
         tf.initialize_all_variables().run()
+        # Run convert_ops to achieve the gradietns converting
+        sess.run(convert_ops)
         # Fetch params to validate initial values
```
separate run~

DjangoPeng commented on tensorflow/python/training/optimizer_test.py:
```diff
@@ -128,10 +127,11 @@ def testGradientsAsVariables(self):
 
         tf.initialize_all_variables().run()
+        # Run convert_ops to achieve the gradietns converting
+        sess.run(convert_ops)
```
Separately run `convert_ops` and `opt_op`!~

DjangoPeng commented:
@mrry I have removed the unnecessary import and `control_dependencies`

mrry commented:
Great, thanks! This LGTM after the tests pass.

mrry commented:
@tensorflow-jenkins test this please.

DjangoPeng commented:
@mrry Sounds good~BTW, what is LGTM :)

mrry commented:
It means "Looks Good To Me." And now the tests have passed, so we can merge it, thanks!
